src/sentry/static/sentry/app/views/groupEvents.jsx

```diff
@@ -10,42 +11,90 @@ import Pagination from '../components/pagination';
 import SearchBar from '../components/searchBar';
 import EventsTable from '../components/eventsTable/eventsTable';
 import {t} from '../locale';
+import withEnvironment from '../utils/withEnvironment';
+import {getQueryEnvironment, getQueryStringWithEnvironment} from '../utils/queryString';
+import EnvironmentStore from '../stores/environmentStore';
+import {setActiveEnvironment} from '../actionCreators/environments';
 
 const GroupEvents = createReactClass({
   displayName: 'GroupEvents',
+
+  propTypes: {
+    environment: PropTypes.object,
+  },
+
   mixins: [ApiMixin, GroupState],
 
   getInitialState() {
-    let queryParams = this.props.location.query;
-    return {
+    const queryParams = this.props.location.query;
+
+    const initialState = {
       eventList: [],
       loading: true,
       error: false,
       pageLinks: '',
       query: queryParams.query || '',
     };
+
+    // If an environment is specified in the query, update the global environment
+    // Otherwise if a global environment is present update the query
+    const queryEnvironment = EnvironmentStore.getByName(
+      getQueryEnvironment(queryParams.query || '')
+    );
+
+    if (queryEnvironment) {
+      setActiveEnvironment(queryEnvironment);
+    } else if (this.props.environment) {
+      initialState.query = getQueryStringWithEnvironment(
+        initialState.query,
+        this.props.environment.name
+      );
+    }
+
+    return initialState;
   },
 
   componentWillMount() {
     this.fetchData();
   },
 
   componentWillReceiveProps(nextProps) {
-    if (
-      nextProps.params.groupId !== this.props.params.groupId ||
-      nextProps.location.search !== this.props.location.search
-    ) {
-      let queryParams = nextProps.location.query;
+    // If query has changed, update the environment with the query environment
+    if (nextProps.location.search !== this.props.location.search) {
+      const queryParams = nextProps.location.query;
+
+      const queryEnvironment = EnvironmentStore.getByName(
+        getQueryEnvironment(queryParams.query || '')
+      );
+
+      if (queryEnvironment) {
+        setActiveEnvironment(queryEnvironment);
+      }
+
       this.setState(
         {
           query: queryParams.query,
         },
         this.fetchData
       );
     }
+
+    // If environment has changed, update query with new environment
+    if (nextProps.environment !== this.props.environment) {
+      const newQueryString = getQueryStringWithEnvironment(
+        nextProps.location.query.query || '',
+        nextProps.environment ? nextProps.environment.name : null
+      );
+      this.setState(
+        {
+          query: newQueryString,
+        },
+        this.handleSearch(newQueryString)
```

Comment:
Why `setState` here if `handleSearch` is going to navigate to a diff route. Also why call `handleSearch` in callback if it doesn't depend on `state`?

Reply:
Good catch, that's not necessary